sci-wcalc/coupled_microstrip_calc.man
.\" $Id: coupled_microstrip_calc.man,v 1.1 2001/11/04 18:30:26 dan Exp $
.\"
.\" Copyright (c) 2001, 2004 Dan McMahill
.\" All rights reserved.
.\"
.\" This code is derived from software written by Dan McMahill
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and.\"or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"        This product includes software developed by Dan McMahill
.\"  4. The name of the author may not be used to endorse or promote products
.\"     derived from this software without specific prior written permission.
.\" 
.\"  THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\"  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\"  OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\"  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\"  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
.\"  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\"  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
.\"  AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\"  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\"  OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\"  SUCH DAMAGE.
.\"

.TH coupled_microstrip_calc 1 "March 2001" "Dan McMahill" "Wcalc"
.\".so ../sci.an
.SH NAME
coupled_microstrip_calc - Coupled microstrip transmission line analysis
.SH CALLING SEQUENCE
.nf
[z0,k,z0e,z0o,kev,kodd] = 
  coupled_microstrip_calc(w,s,h,l,tmet,rho,rough,er,tand,f);
.fi
.SH PARAMETERS
.TP 10
w
: width of coupled microstrip lines (meters)
.TP
s
: spacing between coupled microstrip lines (meters)
.TP
h
: substrate thickness (meters)
.TP
l
: lentgh of coupled microstrip line (meters)
.TP
tmet
: metal thickness (meters)
.TP
rho
: resistivity of the metal (ohm-meters)
.TP
rough
: conductor surface roughness (meters-rms)
.TP
er
: substrate relative permitivitty
.TP
tand
: substrate loss tangent
.TP
f
: analysis frequency (Hz)
.TP
z0
: calculated characteristic impedance (Ohms)
.TP
k
: calculated coupling coefficient
.TP
z0e
: calculated even mode characteristic impedance (Ohms)
.TP
z0o
: calculated odd mode characteristic impedance (Ohms)
.TP
kev
: calculated even mode effective relative permitivitty
.TP
kev
: calculated odd mode effective relative permitivitty
.TP
loss
: calculated loss of the line (dB)
.TP
deltal
: calculated open circuit end correction (meters)
.SH DESCRIPTION
function to analyze the electrical characteristics of a 
coupled microstrip transmission line. The physical
dimensions of the coupled microstrip line are given as inputs and the
characteristic impedance, effective dielectric constant, loss and
open circuit end correction are calculated.

The model accounts for dispersion (frequency dependent propagation
velocity).
.nf

              |<--W-->|<-S->|<--W-->|
               _______       -------
              | metal |     | metal | <- rho,rough
    ----------------------------------------------
   (  dielectric                         /|\\     (
    )   er,tand                       H   |       )
   (                                     \\|/     (
    ----------------------------------------------
    XXXXXXXXXXXXXXXXXX ground XXXXXXXXXXXXXXXXXXXX

.fi
.SH EXAMPLE
.nf
sf=25.4e-6;       // mils to meters conversion factor
w=20*sf;          // width 
s=8*sf;           // spacing
h=15*sf;          // height
l=1100*sf;        // length
tmet=1.4*sf;      // metal thickness
rho=1;            // resitivity relative to copper
rough=0.05*sf;    // metal surface roughness
er=4.5;           // relative dielectric constant
tand=0.01;        // loss tangent
f=2.4e9;          // frequency
[z0,k,z0e,z0o,kev,kodd] = ...
  coupled_microstrip_calc(w,s,h,l,tmet,rho,rough,er,tand,f);
disp(sprintf('z0=%g Ohms, k=%g',z0,k));
disp(sprintf('z0e=%g Ohms, z0o=%g Ohms',z0e,z0o));
disp(sprintf('kev=%g, kodd=%g',kev,kodd));
.fi
.SH SEE ALSO
coupled_microstrip_syn
.SH AUTHOR
Dan McMahill


NEWS
$Id: NEWS,v 1.6 2008/11/30 00:21:16 dan Exp $

Wcalc -- history of user-visible changes. 

Copyright (c) 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005
2006, 2007, 2008, 2009 Dan McMahill

Please report bugs on the sourceforge project page.

===============
Version 1.1
===============
Added self and mutual inductance of two rectangular bars.

Added coplanar waveguide model.

Added coupled stripline model.

Corrected the Q calculations for low frequency (not skindepth) region
in air core inductors.  Also corrected Q calculation for conductors
other than copper.

When compiling with gtk-2.10 or newer, use gtkPrint for printing.
This gives a more professional/standard print dialog and also enables
printing under windows.

Added a preliminary Dutch translation.

Converted the figures to all use solid grayscale fills instead of
pattern fills as the former prints better.

Several updates to the build system to moderize it a bit.  Uses
AM_CONDITIONALS more appropriately, uses AC_PROG_LIBTOOL instead of
AM_PROG_LIBTOOL, etc.

Update the win32 build script to generate an installer that works
correctly under vista.

Add .wc file associations under vista.

Update the octave build infrastructure to work with octave-3.0 and
newer where mex support is built into mkoctfile instead of a
standalone mex shell script.

Switch to GPL licensing.

Create several utility functions to make it faster to build new gtk
forms.

Improved autogen.sh to check versions of several of the auto* tools to
help make it easier to diagnose any issues which may come up.

Repair compilation with the SunPRO compilers.


===============
Version 1.0
===============
Replaced the alert dialog box in the gtk version with a log
window.  This prevents a problem when many alerts are generated
and the user is presented with many many alert windows.

Added win32 support along with a build script for a windows installer.

Fixed a bug which gave lots of error/warning messages in the cgi
frontend for coupled microstrip.

===============
Version 0.9
===============
First public release.

===============
Version 0.1 
===============
Initial tagged version.  This does good analysis and synthesis
of microstrip.  Considered stable for this purpose.  However,
open/save/print and units other than mils are not supported yet.



sci-wcalc/coupled_microstrip_calc.c
/* $Id: coupled_microstrip_calc.c,v 1.1 2001/11/04 18:30:26 dan Exp $ */

static char vcid[] = "$Id$";

/*
 * Copyright (c) 2001 Dan McMahill
 * All rights reserved.
 *
 * This code is derived from software written by Dan McMahill
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *        This product includes software developed Dan McMahill
 * 4. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>

#include "coupled_microstrip.h"
#include "physconst.h"

#include "mex.h"

/*
 * function [z0,k,z0e,z0o,kev,kodd] = 
 *       coupled_microstrip_calc(w,s,h,l,tmet,rho,rough,er,tand,f);
 */

/* Input Arguments */

#define	W_IN	  prhs[0]
#define	S_IN	  prhs[1]
#define	H_IN	  prhs[2]
#define	L_IN	  prhs[3]
#define	TMET_IN   prhs[4]
#define	RHO_IN    prhs[5]
#define	ROUGH_IN  prhs[6]
#define	ER_IN	  prhs[7]
#define	TAND_IN	  prhs[8]
#define	FREQ_IN	  prhs[9]

/* Output Arguments */

#define	Z0_OUT	   plhs[0]
#define	K_OUT	   plhs[1]
#define	Z0E_OUT	   plhs[2]
#define	Z0O_OUT	   plhs[3]
#define	KEV_OUT    plhs[4]
#define	KODD_OUT   plhs[5]


#define CHECK_INPUT(x,y,z,v)                                          \
m = mxGetM(x);                                                        \
n = mxGetN(x);                                                        \
if (!mxIsNumeric(x) ||                                                \
     mxIsComplex(x) ||                                                \
     mxIsSparse(x)  ||                                                \
    !mxIsDouble(x) ) {                                                \
    mexErrMsgTxt(#y " must be a real valued vector "                  \
                    "in COUPLED_MICROSTRIP_CALC.");                   \
}                                                                     \
if (vector && (m>1 || n>1) &&                                         \
    ( (m != rows) || (n != cols) ) ){                                 \
    mexErrMsgTxt(#y ": all vector/matrix inputs must have the same"   \
                    " dimensions in COUPLED_MICROSTRIP_CALC.");       \
    (z) = NULL;                                                       \
}                                                                     \
else if(m*n > 1) {                                                    \
 vector=1;                                                            \
 rows=m;                                                              \
 cols=n;                                                              \
 (z) = &ind;                                                          \
}                                                                     \
else {                                                                \
 (z) = &fixed;                                                        \
}                                                                     \
(v) = mxGetPr(x);                                                     \


/*
 * Note that the V4_COMPAT is for compiling with Matlab.
 * For Scilab, we always want to avoid the `const' part.
 */
void mexFunction(
                 int nlhs,
		 mxArray *plhs[],
                 int nrhs,
#if !defined(V4_COMPAT) && !defined(SCI_MEX)
		 const mxArray *prhs[]
#else
		 mxArray *prhs[]
#endif
		 )
{
  /* inputs */
  double *w,*s,*h,*l,*tmet,*rho,*rough,*er,*tand,*freq;
  unsigned int *ind_w,*ind_s,*ind_h,*ind_l,*ind_tmet,*ind_rho;
  unsigned int *ind_rough,*ind_er,*ind_tand,*ind_freq;

  /* outputs */
  double	*z0,*k,*z0e,*z0o,*kev,*kodd;

  /* number of rows and columns */
  unsigned int rows=1,cols=1;

  /* temp variables */
  unsigned int	m,n;

  /* do we have any vectors in our input */
  int vector=0;

  /* indices for the main loop */
  unsigned int ind=0, fixed=0;

  /* our coupled_microstrip for calculations */
  coupled_microstrip_line *line;


  /*
   * Arguemnt checking and processing
   */

  if (nrhs == 0) {
    mexPrintf("%s\r\n",vcid);
    return;
  }

  /* Check for proper number of arguments */
  if (nrhs != 10) {
    mexErrMsgTxt("wrong number of input arguments to COUPLED_MICROSTRIP_CALC"
		 " (needs 10).");
  } 

  if (nlhs > 4) {
    mexErrMsgTxt("wrong number of output arguments to COUPLED_MICROSTRIP_CALC"
		 " (needs <= 4).");
  }
  
  
  /* 
   * Check the dimensions of the inputs and assign pointers to 
   * the various parameters 
   *
   * CHECK_INPUT( input_variable,
   *              variable_name (for messages),
   *              index,
   *              pointer)
   */
  CHECK_INPUT(W_IN, W, ind_w, w);
  CHECK_INPUT(S_IN, S, ind_s, s);
  CHECK_INPUT(H_IN, H, ind_h, h);
  CHECK_INPUT(L_IN, L, ind_l, l);
  CHECK_INPUT(TMET_IN, TMET, ind_tmet, tmet);
  CHECK_INPUT(RHO_IN, RHO, ind_rho, rho);
  CHECK_INPUT(ROUGH_IN, ROUGH, ind_rough, rough);
  CHECK_INPUT(ER_IN, ER, ind_er, er);
  CHECK_INPUT(TAND_IN, TAND, ind_tand, tand);
  CHECK_INPUT(FREQ_IN, FREQ, ind_freq, freq);

  /* Create matrices for the return arguments */
  Z0_OUT     = mxCreateDoubleMatrix(rows, cols, mxREAL);
  K_OUT      = mxCreateDoubleMatrix(rows, cols, mxREAL);
  Z0E_OUT    = mxCreateDoubleMatrix(rows, cols, mxREAL);
  Z0O_OUT    = mxCreateDoubleMatrix(rows, cols, mxREAL);
  KEV_OUT    = mxCreateDoubleMatrix(rows, cols, mxREAL);
  KODD_OUT   = mxCreateDoubleMatrix(rows, cols, mxREAL);
  
  /* output pointers */
  z0     = mxGetPr(Z0_OUT);
  k      = mxGetPr(K_OUT);
  z0e    = mxGetPr(Z0E_OUT);
  z0o    = mxGetPr(Z0O_OUT);
  kev    = mxGetPr(KEV_OUT);
  kodd   = mxGetPr(KODD_OUT);

  /* the actual computation */
  line = coupled_microstrip_line_new();

  for (ind=0; ind<(rows*cols); ind++){
    /*
     * copy over the parameters.  Note, some of the ind_* pointers
     * point to 'ind' and others point to 'fixed'.  This lets us pick
     * which inputs are scalar and which are matrices easily.  (It's
     * done when processsing the input arguments.
     */
    line->w           = w[*ind_w];
    line->s           = s[*ind_s];
    line->l           = l[*ind_l];
    line->subs->h     = h[*ind_h];
    line->subs->tmet  = tmet[*ind_tmet];
    line->subs->rho   = rho[*ind_rho];
    line->subs->rough = rough[*ind_rough];
    line->subs->er    = er[*ind_er];
    line->subs->tand  = tand[*ind_tand];
    line->freq        = freq[*ind_freq];

    /* run the calculation */
    coupled_microstrip_calc(line,line->freq);

    /* extract the outputs */
    z0[ind]     = line->z0;
    k[ind]      = line->k;
    z0e[ind]    = line->z0e;
    z0o[ind]    = line->z0o;
    kev[ind]    = line->kev;
    kodd[ind]   = line->kodd;
  }

  /* clean up */
  coupled_microstrip_line_free(line);

  return;
}
